embedded/store/immustore.go

```diff
@@ -1425,31 +1476,152 @@ func (s *ImmuStore) performPreCommit(tx *Tx, ts int64, blTxID uint64) error {
 
 	s.precommitWHub.DoneUpto(s.preCommittedTxID)
 
-	var cb [cLogEntrySize]byte
-	binary.BigEndian.PutUint64(cb[:], uint64(txOff))
-	binary.BigEndian.PutUint32(cb[offsetSize:], uint32(txSize))
+	err = s.cLogBuf.put(s.preCommittedTxID, alh, txOff, txSize)
+	if err != nil {
+		return err
+	}
+
+	if !s.synced {
+		s.durablePrecommitWHub.DoneUpto(s.preCommittedTxID)
+
+		return s.mayCommit()
+	}
+
+	return nil
+}
+
+func (s *ImmuStore) EnableExternalCommitAllowance() error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.closed {
+		return ErrAlreadyClosed
+	}
+
+	s.commitStateRWMutex.Lock()
+	defer s.commitStateRWMutex.Unlock()
+
+	s.useExternalCommitAllowance = true
+	s.commitAllowedUpToTxID = s.committedTxID
+
+	return nil
+}
+
+func (s *ImmuStore) DisableExternalCommitAllowance() error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.closed {
+		return ErrAlreadyClosed
+	}
+
+	s.commitStateRWMutex.Lock()
+	defer s.commitStateRWMutex.Unlock()
+
+	s.useExternalCommitAllowance = false
+
+	return nil
+}
 
-	if s.synced {
-		copy(s.cLogBuf[int(s.preCommittedTxID-s.committedTxID-1)*cLogEntrySize:], cb[:])
+func (s *ImmuStore) AllowCommitUpto(txID uint64) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.closed {
+		return ErrAlreadyClosed
+	}
+
+	if !s.useExternalCommitAllowance {
+		return fmt.Errorf("%w: the external commit allowance mode is not enabled", ErrIllegalState)
+	}
+
+	s.commitStateRWMutex.Lock()
+	defer s.commitStateRWMutex.Unlock()
+
+	if txID <= s.commitAllowedUpToTxID {
+		// once a commit is allowed, it cannot be revoked
+		return nil
+	}
+
+	if s.preCommittedTxID < txID {
+		// commit allowances apply only to pre-committed transactions
```

Comment:
hm, no sure if this could lead to an inconsistent situation

Reply:
This would not intruduce inconsistency but could block the replica from performing the commit, it would automatically fix in the future but that requires another commit from the primary with updated information up to what TX the commit can be performed.